Updated parseEntityTransformParams to handle keys with '.' in them

webark opened the pull request:
Signed-off-by: Mark David Avery <mark@webark.cc>

## Hey, I just made a Pull Request!

When writing queries against the entity list, this will enable using keys with `.` in them. This includes annotations such as `backstage.io/source-location` and the like.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

webark commented:
Applicable thread.

https://discord.com/channels/687207715902193673/968415863487225866

freben commented on plugins/catalog-backend/src/service/request/parseEntityTransformParams.ts:
```diff
@@ -20,6 +20,22 @@ import lodash from 'lodash';
+  }
+
+  return pathArray;
+}
```
This has bugs - the test succeeds "by accident" since it'll stop at the annotations level and return that, and the example input data there only had one annotation. I think the problem is, your currentPathPart doesn't add a dot between parts.

While you are in this function, it could actually have kept track of the current level of the hierarchy and returned the value it found straight away (so the return type could be `[path: string[], value: unknown]` or so). That'd probably be significantly more efficient than constantly doing `lodash.has` and re-traversing down the hierarchy over and over at every level.

webark replied:
Ok. I'll make that change.

webark replied:
@freben updated to fix the bug and add more tests. 

Will come back and change the implementation to what you suggested, just wanted to first correct this implementation and add the additional tests.

webark commented:
@freben so the two implimenations that I setted on where

```typescript
function getPathArrayAndValue(input: Entity, field: string) {
  const pathArray: string[] = [];
  let inputSubset = input as any;

  field.split('.').forEach((pathPart, index, fieldParts) => {
    if (Object.hasOwn(inputSubset, pathPart)) {
      pathArray.push(pathPart);
      inputSubset = inputSubset[pathPart];
    } else if (fieldParts[index + 1]) {
      fieldParts[index + 1] = `${pathPart}.${fieldParts[index + 1]}`;
    }
  });

  return [pathArray, inputSubset];
}
```
and 
```typescript
function getPathArray(input: Entity, field: string) {
  const pathArray: string[] = [];

  field.split('.').forEach((pathPart, index, fieldParts) => {
    if (lodash.has(input, pathArray.concat(pathPart))) {
      pathArray.push(pathPart);
    } else if (fieldParts[index + 1]) {
      fieldParts[index + 1] = `${pathPart}.${fieldParts[index + 1]}`;
    }
  });

  return pathArray;
}
```

My one concern is that we arent using lodash to retrieve or check (with _.has) the key, but we will be using it to set it later on, and if that could be an area for odd bugs to arise.

I'm open to your call and feedback @freben . thanks for taking the time! 😊

webark commented:
I ended up just pushing the `getPathArrayAndValue` option since that was what you were suggesting. If we want to move back to the `lodash.has` approach, I'm fine with that and can update it as such. Thanks!

webark commented:
Oh, `hasOwn` was introduced in 16, and we're still supporting 14. Makes sense. I'll update thatt.

webark commented:
OK, I fixed that commit that wasn't signed.

Let me know if you want me to squash the commits, or if that's something that you will do when you merge/rebase in the PR.

freben commented on plugins/catalog-backend/src/service/request/parseEntityTransformParams.ts:
```diff
@@ -20,6 +20,22 @@ import lodash from 'lodash';
+function getPathArrayAndValue(input: Entity, field: string) {
+  return field.split('.').reduce(
+    ([pathArray, inputSubset], pathPart, index, fieldParts) => {
+      if (inputSubset.hasOwnProperty(pathPart)) {
```
I don't think you can do this in a `reduce` easily - before even going into the `if` statements you need to check basically that `typeof inputSubset === 'object'` or similar, and if it isn't, bail out early and return undefined - because there is no point in trying to descend into a null or a number or so. So this should probably be a for..of loop instead over the parts, or similar

freben replied:
You could write a test that tries to fetch `spec.type`, and then a test that tries to fetch `spec.type.doesnotexist`

webark replied:
Ohh sure. I was using a "hasOwn" before, but that was not supported in node 14. 😞 

I'll add that test.

I had a for of loop originally, but it felt to cumbersome. I'll switch it back if that's preferred.

webark replied:
@freben so 
> You could write a test that tries to fetch spec.type, and then a test that tries to fetch spec.type.doesnotexist

`hasOwnProperty` exists on all javascript types except null and undefined. So I added some tests before I thought of the null and undefined scenarios, and will update to handle those later.

webark replied:
Ok. I switched it to loash's `hasIn` which seems to be a polyfill for 'hasOwn`. I like this, cause it serves the need of using lodah to check the key, as well as set the key, so there's less likely to be an odd bug between the two.

I kept it as a reduce for now. I could go back to a for of, that was a little like
```
function getPathArray(input: Entity, field: string) {
  const pathArray = [];
  let currentPathPart = '';

  for (const pathPart of field.split('.')) {
    currentPathPart += pathPart;

    if (lodash.has(input, pathArray.concat(currentPathPart))) {
      pathArray.push(currentPathPart);
      currentPathPart = '';
    } else {
      currentPathPart += '.';
    }
  }

  return pathArray;
}
```

Let me know if you want to be to go back to that approach, but with stepping through the object each cycle to get the value.

suuus commented:
Thank you for contributing during Hacktoberfest 🍂! You can claim the Backstage Hacktoberfest Holopin at https://bck.st/hacktoberfest-holopin 🙌🏻
